Skip unavailable replicas in parallel distributed insert select

tavplubix opened the pull request:
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Skip unavailable replicas when executing parallel distributed `INSERT SELECT`

tavplubix committed:
Update StorageDistributed.cpp

robot-clickhouse commented:
<!-- automatic status comment for PR #58931 from ClickHouse/ClickHouse:tavplubix-patch-10 -->
*This is an automated comment for commit d6c0908e7c5d94a77373bea36ac45a34ce4a8341 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58931/d6c0908e7c5d94a77373bea36ac45a34ce4a8341/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58931/d6c0908e7c5d94a77373bea36ac45a34ce4a8341/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

tavplubix commented:
Not sure if we should add another setting (I don't like using `skip_unavailable_shards` for skipping unavailable replicas) or even do it by default
Also, it would be great to add a test

tavplubix commented:
> Not sure if we should add another setting (I don't like using `skip_unavailable_shards` for skipping unavailable replicas) or even do it by default

Although, look like we already use `skip_unavailable_shards` for skipping unavailable replicas in `s3Cluster`

Avogar commented:
> Although, look like we already use skip_unavailable_shards for skipping unavailable replicas in s3Cluster

Well, we explicitely set `skip_unavailable_shards` always to `true` in `*Cluster` functions: https://github.com/ClickHouse/ClickHouse/blob/67f74b552004d846be4f417a5a808676cad283c3/src/Storages/IStorageCluster.cpp#L218-L219
So, we if user changes `skip_unavailable_shards` it doesn't affect  `*Cluster` functions. Also, I think we can change the code in `IStorageCluster` and don`t copy settings with changed `skip_unavailable_shards=true` but just use optional `skip_unavailable_endpoints` argument of `ConnectionPoolWithFailover::getMany`.

tavplubix commented:
> But I guess we still use skip_unavailable_shards as skip unavailable replicas in clusterAllReplicas

It's fine because `clusterAllReplicas` means "treat each replica as a separate shard"

tavplubix committed:
Update StorageDistributed.cpp

tavplubix committed:
Update StorageDistributed.cpp

tavplubix committed:
add a test

tavplubix committed:
Update 01099_parallel_distributed_insert_select.sql

tavplubix commented:
Integration tests (tsan) [3/6] - test_parallel_replicas_custom_key_failover was broken in master, reverted
Performance Comparison [1/4] - https://github.com/ClickHouse/ClickHouse/pull/59070
Stateful tests (tsan, ParallelReplicas) - 00165_jit_aggregate_functions is flaky (probably too long for tsan)
Stateless tests (release, analyzer) - 02901_parallel_replicas_rollup was broken in master, reverted
Stress test (msan) - https://github.com/ClickHouse/ClickHouse/issues/58509

devcrafter commented on src/Storages/StorageDistributed.cpp:
```diff
@@ -1140,22 +1140,20 @@ std::optional<QueryPipeline> StorageDistributed::distributedWriteFromClusterStor
+    for (const auto & replicas : getCluster()->getShardsInfo())
     {
+        /// Skip unavailable hosts if necessary
+        auto try_results = replicas.pool->getMany(timeouts, current_settings, PoolMode::GET_MANY, /*async_callback*/ {}, /*skip_unavailable_endpoints*/ true);
```
If the comment on line 1152 is correct. Then `PoolMode::GET_ONE` is sufficient here

tavplubix replied:
I guess the comment is either incorrect or misleading (because we get the cluster from `StorageDistributed::getCluster()`), we can ask @nikitamikhaylov

tavplubix replied:
Yes, it's copy-paste from https://github.com/ClickHouse/ClickHouse/blob/4315d44517e1ca0412c7dc7c6c39dc1c17881002/src/Storages/StorageReplicatedMergeTree.cpp#L5562-L5565

And looks like we have exactly the same issue in `StorageReplicatedMergeTree::distributedWriteFromClusterStorage` (and that's why we should avoid copy-paste when possible)
